src/TierRegistry.sol

```diff
@@ -0,0 +1,334 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity 0.8.21;
+
+import {AggregatorV3Interface} from "chainlink/v0.8/interfaces/AggregatorV3Interface.sol";
+import {AccessControlEnumerable} from "openzeppelin/contracts/access/AccessControlEnumerable.sol";
+import {Pausable} from "openzeppelin/contracts/security/Pausable.sol";
+import {FixedPointMathLib} from "solmate/src/utils/FixedPointMathLib.sol";
+import "openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import {ITierRegistry} from "./interfaces/ITierRegistry.sol";
+import {TransferHelper} from "./libraries/TransferHelper.sol";
+
+/**
+ * @title Farcaster StorageRegistry
+ *
+ * @notice See https://github.com/farcasterxyz/contracts/blob/v3.1.0/docs/docs.md for an overview.
+ *
+ * @custom:security-contact security@merklemanufactory.com
+ */
+contract TierRegistry is ITierRegistry, AccessControlEnumerable, Pausable {
+    using FixedPointMathLib for uint256;
+    using TransferHelper for address;
+    using SafeERC20 for IERC20;
+
+    /*//////////////////////////////////////////////////////////////
+                                 ERRORS
+    //////////////////////////////////////////////////////////////*/
+
+    /// @dev Revert if the caller attempts to rent storage after the contract is deprecated.
+    error ContractDeprecated();
+
+    /// @dev Revert if the caller attempts to rent zero units.
```

Comment:
will avoid leaving comments on comments, but let's make sure we go through and do a final pass at some point.